tactics/auto.ml

```diff
@@ -258,50 +262,56 @@ let tclTRY_dbg d tac =
 let auto_flags_of_state st =
   auto_unif_flags_of TransparentState.full st
 
-let hintmap_of env sigma secvars hdc concl =
-  match hdc with
-  | None -> Hint_db.map_none ~secvars
-  | Some hdc ->
-     if occur_existential sigma concl then
-       (fun db -> match Hint_db.map_eauto env sigma ~secvars hdc concl db with
-                  | ModeMatch (_, l) -> l
-                  | ModeMismatch -> [])
-     else Hint_db.map_auto env sigma ~secvars hdc concl
+let hintmap_of env sigma secvars concl =
+  try
+    let hdc = decompose_app_bound sigma concl in
```

Comment:
You should move out the call to `decompose_app_bound` otherwise it is repeated in the List.map_append below.

Reply:
(Basically I'm arguing for a partial revert of the part of this patch where hdc is not passed as an argument. This matters for efficiency, and you can probably write it in a more compact way still if you're afraid of code duplication.)

Reply:
I intentionally compute `hdc` from `env`, `sigma`, `secvars`, `concl` (not `db`) and then return a function `let hintmap = hintmap_of env sigma secvars concl in ...`, which is applied to each `db` separately with one precomputed `hdc`. I did assume that OCaml does the proper optimization and does not reevaluate it each time in `map_append`.
Are you sure that OCaml does it the slow way? If so, then of course I would precompute `hdc` externally and pass it as a redundant argument.

Reply:
By the way, the same I thought to be true for the computation of `Termops.occur_existential sigma concl`, which otherwise would be unnecessarily reevaluated in `List.map_append`.

Reply:
Anyway, I now reverted to the previous `hdc` precomputation to be on the safe side, as you suggest.